tool/tsh/proxy.go

```diff
@@ -145,94 +168,141 @@ func setupJumpHost(cf *CLIConf, tc *libclient.TeleportClient, sp sshProxyParams)
+	client, err := makeSSHClient(ctx, upstreamConn, upstreamConn.RemoteAddr().String(), &ssh.ClientConfig{
+		User: tc.HostLogin,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeysCallback(tc.LocalAgent().Signers),
+		},
+		HostKeyCallback: tc.HostKeyCallback,
+	})
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	tlsConfig := &tls.Config{
-		RootCAs: pool,
+	defer client.Close()
+
+	sess, err := client.NewSession(ctx)
+	if err != nil {
+		return trace.Wrap(err)
 	}
+	defer sess.Close()
 
-	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
-		RemoteProxyAddr:    net.JoinHostPort(params.proxyHost, params.proxyPort),
-		Protocols:          []alpncommon.Protocol{alpncommon.ProtocolProxySSH},
-		InsecureSkipVerify: cf.InsecureSkipVerify,
-		ParentContext:      cf.Context,
-		SNI:                params.proxyHost,
-		SSHUser:            tc.HostLogin,
-		SSHUserHost:        fmt.Sprintf("%s:%s", params.targetHost, params.targetPort),
-		SSHHostKeyCallback: tc.HostKeyCallback,
-		SSHTrustedCluster:  params.clusterName,
-		ClientTLSConfig:    tlsConfig,
-	})
+	err = agent.ForwardToAgent(client.Client, tc.LocalAgent())
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	defer lp.Close()
-	if err := lp.SSHProxy(cf.Context, tc.LocalAgent()); err != nil {
+	err = agent.RequestAgentForwarding(sess)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	sshUserHost := fmt.Sprintf("%s:%s", sp.targetHost, sp.targetPort)
+	if err = sess.RequestSubsystem(proxySubsystemName(sshUserHost, sp.clusterName)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := proxySession(ctx, sess); err != nil {
 		return trace.Wrap(err)
 	}
 	return nil
 }
 
-func sshProxy(tc *libclient.TeleportClient, params sshProxyParams) error {
-	sshPath, err := getSSHPath()
+func dialSSHProxy(ctx context.Context, tc *libclient.TeleportClient, sp sshProxyParams) (net.Conn, error) {
+	remoteProxyAddr := net.JoinHostPort(sp.proxyHost, sp.proxyPort)
+
+	if !sp.tlsRouting {
+		conn, err := net.Dial("tcp", remoteProxyAddr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return conn, nil
+	}
+
+	pool, err := tc.LocalAgent().ClientCertPool(sp.clusterName)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs:            pool,
+		NextProtos:         []string{string(alpncommon.ProtocolProxySSH)},
+		InsecureSkipVerify: tc.InsecureSkipVerify,
+		ServerName:         sp.proxyHost,
 	}
-	keysDir := profile.FullProfilePath(tc.Config.KeysDir)
-	knownHostsPath := keypaths.KnownHostsPath(keysDir)
 
-	args := []string{
-		"-A",
-		// TODO: remove once we finish deprecating SHA1
-		"-o", fmt.Sprintf("PubkeyAcceptedKeyTypes=+%s", ssh.CertAlgoRSAv01),
-		"-o", fmt.Sprintf("UserKnownHostsFile=%s", knownHostsPath),
-		"-p", params.proxyPort,
-		params.proxyHost,
-		"-s",
-		fmt.Sprintf("proxy:%s:%s@%s", params.targetHost, params.targetPort, params.clusterName),
+	conn, err := tls.Dial("tcp", remoteProxyAddr, tlsConfig)
```

Comment:
The same as above, since we have a context in this function we should make use of it
```suggestion
	conn, err := (&tls.Dialer{Config: tlsConfig}).DialContext(ctx, "tcp", remoteProxyAddr)
```